statsd/buffer.go

```diff
@@ -39,22 +39,28 @@ func newStatsdBuffer(maxSize, maxElements int) *statsdBuffer {
 	}
 }
 
-func (b *statsdBuffer) writeGauge(namespace string, globalTags []string, name string, value float64, tags []string, rate float64) error {
+func (b *statsdBuffer) writeGauge(namespace string, globalTags []string, name string, value float64, tags []string, rate float64, timestamp int64) error {
 	if b.elementCount >= b.maxElements {
 		return errBufferFull
 	}
 	originalBuffer := b.buffer
 	b.buffer = appendGauge(b.buffer, namespace, globalTags, name, value, tags, rate)
+	if timestamp != noTimestamp {
```

Comment:
I'm kind of disagreeing with that because to me, it's when writing a gauge and a count (so, in `writeGauge` and `writeCount`) that we don't want to write a zero timestamp. While having `appendTimestamp` writing a `0` value is correct by meaning and should be possible. 
Also, your solution means we have this:
```
	b.buffer = appendGauge(b.buffer, namespace, globalTags, name, value, tags, rate)
	b.buffer = appendTimestamp(b.buffer, timestamp)
```
while no timestamp will be written if `timestamp == 0`, which makes this invisible to the reader.

But honestly, I've hesitated when implementing this and no strong opinion so I'll follow your idea.

Reply:
First I think this would avoid duplicating the logic of `if timestamp != noTimestamp {` in multiple places as well as making it future proof for other devs.

But from a design POV I think the `appendTimestamp` should encapsulate the logic around timestamp. `writeGauge` or `writeCount` should not have to know every implementation details of the protocol. all they know is the order of the building blocks (metric name then type then tags, ...).